sycl/source/detail/scheduler/commands.cpp

```diff
@@ -1797,71 +1797,100 @@ cl_int ExecCGCommand::enqueueImp() {
 
     // Run OpenCL kernel
     sycl::context Context = MQueue->get_context();
-    const detail::plugin &Plugin = MQueue->getPlugin();
     RT::PiKernel Kernel = nullptr;
+    std::mutex *KernelMutex = nullptr;
 
     if (nullptr != ExecKernel->MSyclKernel) {
       assert(ExecKernel->MSyclKernel->get_info<info::kernel::context>() ==
              Context);
       Kernel = ExecKernel->MSyclKernel->getHandleRef();
-    } else
-      Kernel = detail::ProgramManager::getInstance().getOrCreateKernel(
-          ExecKernel->MOSModuleHandle, Context, ExecKernel->MKernelName,
-          nullptr);
 
-    for (ArgDesc &Arg : ExecKernel->MArgs) {
-      switch (Arg.MType) {
-      case kernel_param_kind_t::kind_accessor: {
-        Requirement *Req = (Requirement *)(Arg.MPtr);
-        AllocaCommandBase *AllocaCmd = getAllocaForReq(Req);
-        RT::PiMem MemArg = (RT::PiMem)AllocaCmd->getMemAllocation();
-        if (Plugin.getBackend() == backend::opencl) {
+      auto SyclProg = detail::getSyclObjImpl(
+          ExecKernel->MSyclKernel->get_info<info::kernel::program>());
+      if (SyclProg->is_cacheable()) {
+        RT::PiKernel FoundKernel = nullptr;
+        std::tie(FoundKernel, KernelMutex) =
+            detail::ProgramManager::getInstance().getOrCreateKernel(
+                ExecKernel->MOSModuleHandle,
+                ExecKernel->MSyclKernel->get_info<info::kernel::context>(),
+                ExecKernel->MKernelName, SyclProg.get());
+        assert(FoundKernel == Kernel);
+      }
+    } else {
+      std::tie(Kernel, KernelMutex) =
+          detail::ProgramManager::getInstance().getOrCreateKernel(
+              ExecKernel->MOSModuleHandle, Context, ExecKernel->MKernelName,
+              nullptr);
+    }
+
+    auto SetKernelParamsAndLaunch = [this, &ExecKernel, &Kernel, &NDRDesc,
```

Comment:
Nit:  lambdas are supposed to be short lightweight inline functions, otherwise, like in this case, it rather hinders understanding. This would ideally be a separate function. Plus instead of duplicating the call below
```
    if (KernelMutex != nullptr) {
        // For cacheable kernels, we use per-kernel mutex
        std::lock_guard<std::mutex> Lock(*KernelMutex);
      Error = SetKernelParamsAndLaunch();
    } else {
      Error = SetKernelParamsAndLaunch();
```
mutex could be a parameter, non-null meaning that the lock needs to be acquired

Reply:
Thanks for the comment, this is a good idea: replace lambda with a separate function. Regarding passing a mutex as a parameter,  I believe it is not a good idea because then I don't see how to use `std::lock_guard` without duplication of whole code of the function and using `lock()`, `unlock()` - as discussed above is bad practice. (this is a loss of automatically lock/unlock when the object is constructed/destroyed)